### PR TITLE
ESM Agent - Update

### PR DESCRIPTION
**Pull request by RCantu92**
# ESM Agent

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] New Agent
- [ ] Fix an agent error
- [x] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [ ] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Agent contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [ ] Network calls are properly being cached (if duplicated calls can occur)
- [ ] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Agent can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this agent please describe it here.
- Add functionality to detect when ESM address has been updated on ChainLog contract. (via `UpdateAddress` event emission.
- Add two unit tests in `agent.spect.ts` for above functionality.
- Pass `fetcher` down to "sub-agents" (i.e. individual event handlers)

### CODE REVIEW

File: Maker-Agents/ESMAgent/src/fire.event.ts [hunk 1]
```diff
@@ -20,12 +21,12 @@ const createFindingGenerator = (alertID: string, ESM_address: string, _from: str
     });
 };
 
-const provideESMFireEventAgent = (_alertID: string, _contractAddress: string): HandleTransaction => {
+const provideESMFireEventAgent = (_alertID: string, fetcher: AddressFetcher): HandleTransaction => {
   return async (txEvent: TransactionEvent): Promise<Finding[]> => {
     const agentHandler = provideEventCheckerHandler(
-      createFindingGenerator(_alertID, _contractAddress, txEvent.from),
+      createFindingGenerator(_alertID, await fetcher.esmAddress, txEvent.from),
```
Comment (yasmine9) [inline Maker-Agents/ESMAgent/src/fire.event.ts:27]: No need for the await statement here

**Review by yasmine9 — APPROVED** [review]
One minor comment. The rest looks good to me

File: Maker-Agents/ESMAgent/src/join.event.ts [hunk 2]
```diff
@@ -39,12 +40,12 @@ const createFindingGenerator = (_alertID: string): FindingGenerator => {
     });
 };
```
Comment (0xtaf) [inline Maker-Agents/ESMAgent/src/join.event.ts:41]: in filterLog function, why is there unused index and array parameters? how or where are they used?

File: Maker-Agents/ESMAgent/src/address.fetcher.spec.ts [hunk 3]
```diff
@@ -1,28 +1,33 @@
...
+  }
+
+  beforeEach(() => mockProvider.clear());
 
   it("should store correct MCD_ESM address", async () => {
-    initialize();
-    await fetcher.getEsmAddress(1);
-    expect(fetcher.esmAddress).toStrictEqual(ESM_CONTRACT);
+    for (let [esmAddress, blockNumber] of TEST_CASES) {
+      const fetcher: AddressFetcher = new AddressFetcher(mockProvider as any, CHAINLOG_ADDRESS);
```
Comment (yasmine9) [inline Maker-Agents/ESMAgent/src/address.fetcher.spec.ts:26]: Per Lazaro's note, no need to initiate the fetcher every time, this line can be moved to Line 14

**Review by yasmine9 — CHANGES_REQUESTED** [review]

File: Maker-Agents/ESMAgent/src/join.event.ts (hunk 2 above)
Comment (RCantu92) [reply Maker-Agents/ESMAgent/src/join.event.ts:41]: The `filter` argument for the `provideEventCheckerHandler` from the `forta-agent-tools` needs to be of that type. [Here](https://github.com/NethermindEth/general-agents-module#--event-checker-handler) is the description for it.

Also, for what I need, the `log` argument will suffice.

**Review by yasmine9 — APPROVED** [review]

File: Maker-Agents/ESMAgent/src/address.fetcher.spec.ts [hunk 4]
```diff
@@ -1,28 +1,33 @@
...
 
   it("should store correct MCD_ESM address", async () => {
-    initialize();
-    await fetcher.getEsmAddress(1);
-    expect(fetcher.esmAddress).toStrictEqual(ESM_CONTRACT);
+    for (let [esmAddress, blockNumber] of TEST_CASES) {
+      createGetAddressCall(blockNumber, esmAddress);
+
+      const fetchedEsmAddres: string = await fetcher.getEsmAddress(blockNumber);
+      expect(fetchedEsmAddres).toStrictEqual(esmAddress);
```
Comment (0xtaf) [inline Maker-Agents/ESMAgent/src/address.fetcher.spec.ts:30]: typo here

**Review by 0xtaf — APPROVED** [review]

**Pull request merged**